### PR TITLE
Python CLI PR1

### PR DESCRIPTION
**Pull request by agentmilindu**
Python CLI for Stratos 

*( Do not merge )*

### CODE REVIEW

File: components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+class CLI(Cmd):
+    """Apache Stratos CLI"""
+
+    prompt = Configs.stratos_prompt
+    # resolving the '-' issue
+    Cmd.legalChars = '-' + Cmd.legalChars
+
+    def __init__(self):
+        # resolving the '-' issue
+        [Cmd.shortcuts.update({a[3:].replace('_', '-'): a[3:]}) for a in self.get_names() if a.startswith('do_')]
```
Comment (chamilad) [inline components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py]: Code readability is lesser here. IMO we should go for a cleaner approach with self describing code.

File: components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            table.print_table()
+
+    @options([])
+    def do_deploy_user(self, line , opts=None):
+        """Illustrate the base class method use."""
+        print("hello User")
+        try:
+            Stratos.deploy_user()
+        except ValueError as e:
+            self.perror("sdc")
```
Comment (chamilad) [inline components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py]: New line at the end of file - PEP8

File: components/org.apache.stratos.python.cli/src/main/python/cli/Configs.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+stratos_prompt = "stratos> "
+
+stratos_dir = "~/.stratos"
+log_file_name = "stratos-cli.log"
+
+stratos_dir_path = os.path.expanduser(stratos_dir)
+log_file_path = os.path.join(stratos_dir_path, log_file_name)
+
+stratos_url = os.getenv('STRATOS_URL', "https://localhost:9443/")
```
Comment (chamilad) [inline components/org.apache.stratos.python.cli/src/main/python/cli/Configs.py]: Are these configurable options by the user? If they are there shouldn't be code here. Better go for a config file such as an *.ini file and read from it.

File: components/org.apache.stratos.python.cli/src/main/python/cli/Main.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import sys
```
Comment (chamilad) [inline components/org.apache.stratos.python.cli/src/main/python/cli/Main.py]: License statement missing. You can configure most IDEs to automatically insert the statement to a new file.

File: components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    def do_list_applications(self, line , opts=None):
+        """Illustrate the base class method use."""
+        applications = Stratos.list_applications()
+        if not applications:
+            print("No applications found")
+        else:
+            table = PrintableTable()
+            rows = [["Type", "Category", "Name", "Description", "Version", "Multi-Tenant"]]
+            for application in applications:
+                rows.append([application['type'], application['category'], application['displayName'], application['description'], application['version'], application['multiTenant']])
```
Comment (chamilad) [inline components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py]: check 180 char limit

File: components/org.apache.stratos.python.cli/src/main/python/cli/Stratos.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+
+class Stratos:
+    """Apache Stratos Python API"""
+
+    def __init__(self):
+        pass
+
+    """
+    # Users
```
Comment (chamilad) [inline components/org.apache.stratos.python.cli/src/main/python/cli/Stratos.py]: Include more descriptive comments for both methods and modules/classes.

File: components/org.apache.stratos.python.cli/src/main/python/cli/Utils.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    def __init__(self):
+        Texttable.__init__(self)
+        self.set_deco(Texttable.BORDER | Texttable.HEADER | Texttable.VLINES)
+
+    def print_table(self):
+        print(self.draw())
+
+def auth(func):
+    """Authenticate"""
+    def auth_inner(self, *args, **kwargs):
```
Comment (chamilad) [inline components/org.apache.stratos.python.cli/src/main/python/cli/Utils.py]: Why an inner function here?

File: components/org.apache.stratos.python.cli/src/main/python/cli/Configs.py (hunk 3 above)
Comment (agentmilindu) [inline components/org.apache.stratos.python.cli/src/main/python/cli/Configs.py]: At the moment they are not configurable by the user, but let's try to go for a config file such as an *.ini  as you suggest, once we are done with the rest.

File: components/org.apache.stratos.python.cli/src/main/python/cli/Main.py (hunk 4 above)
Comment (agentmilindu) [inline components/org.apache.stratos.python.cli/src/main/python/cli/Main.py]: Fixed :)

File: components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py (hunk 5 above)
Comment (agentmilindu) [inline components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py]: Fixed :)

File: components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py (hunk 2 above)
Comment (agentmilindu) [inline components/org.apache.stratos.python.cli/src/main/python/cli/CLI.py]: Fixed :)

**Comment by imesh** [comment]
Thanks Milindu for your contribution! This pull request is now merged to stratos-4.1.x branch. It will be released with 4.1.4.

**Pull request closed without merging**